openslides_backend/action/actions/topic/import_.py

```diff
@@ -1,51 +1,120 @@
+from ...mixins.import_mixins import (
+    ImportMixin,
+    ImportRow,
+    ImportState,
+    Lookup,
+    ResultType,
+)
 from ...util.default_schema import DefaultSchema
 from ...util.register import register_action
+from ..agenda_item.update import AgendaItemUpdate
 from .create import TopicCreate
-from .mixins import DuplicateCheckMixin
+from .update import TopicUpdate
 
 
 @register_action("topic.import")
-class TopicImport(DuplicateCheckMixin, ImportMixin):
+class TopicImport(ImportMixin):
     """
     Action to import a result from the import_preview.
     """
 
-    model = ActionWorker()
-    schema = DefaultSchema(ActionWorker()).get_default_schema(
+    model = ImportPreview()
+    schema = DefaultSchema(ImportPreview()).get_default_schema(
         additional_required_fields={
             "id": required_id_schema,
             "import": {"type": "boolean"},
         }
     )
     permission = Permissions.AgendaItem.CAN_MANAGE
     import_name = "topic"
+    agenda_item_fields = ["agenda_comment", "agenda_duration", "agenda_type"]
 
     def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
-        instance = super().update_instance(instance)
-
-        # handle abort in on_success
         if not instance["import"]:
             return {}
 
+        instance = super().update_instance(instance)
+
         meeting_id = self.get_meeting_id(instance)
-        self.init_duplicate_set(meeting_id)
-        action_payload = [
-            entry["data"]
-            for entry in self.result.get("rows", [])
-            if (entry["state"] in (ImportState.NEW, ImportState.WARNING))
-            and not self.check_for_duplicate(entry["data"]["title"])
-        ]
-        self.execute_other_action(TopicCreate, action_payload)
-        self.error = False
-        return instance
+        self.setup_lookups(self.result.get("rows", []), meeting_id)
+
+        self.rows = [self.validate_entry(row) for row in self.result["rows"]]
+
+        if self.import_state != ImportState.ERROR:
+            create_action_payload: List[Dict[str, Any]] = []
+            update_action_payload: List[Dict[str, Any]] = []
+            update_agenda_item_payload: List[Dict[str, Any]] = []
+            self.flatten_object_fields(["title"])
+            for row in self.rows:
+                entry = row["data"]
+                if row["state"] == ImportState.NEW:
+                    create_action_payload.append(entry)
+                else:
+                    agenda_item = {
+                        field[7:]: value
+                        for field in self.agenda_item_fields
+                        if (value := entry.pop(field, None)) is not None
+                    }
+                    if agenda_item:
+                        agenda_item["id"] = self.topic_lookup.get_field_by_name(
+                            entry["title"], "agenda_item_id"
+                        )
+                        update_agenda_item_payload.append(agenda_item)
```

Comment:
Merge this with the code here:
https://github.com/OpenSlides/openslides-backend/blob/95d52ad8fd9782d81e6d4183dc4725086919e13c/openslides_backend/action/actions/agenda_item/agenda_creation.py#L82
Also, magic numbers (like the 7 here) are almost never a good idea

Reply:
done